heron/healthmgr/src/java/org/apache/heron/healthmgr/common/PhysicalPlanProvider.java

```diff
@@ -19,70 +19,135 @@
-        physicalPlan = null;
-      }
-    });
-    eventManager.addEventListener(ContainerRestart.class, new EventHandler<ContainerRestart>() {
-      /**
-       * Invalidates cached physical plan on receiving container restart notification
-       */
-      @Override
-      public synchronized void onEvent(ContainerRestart event) {
-        LOG.info("Received container restart event, invalidating cached PhysicalPlan: "
-            + event.type());
-        physicalPlan = null;
-      }
-    });
+  }
+
+  protected PhysicalPlan ParseResponseToPhysicalPlan(byte[] responseData) {
+    // byte to base64 string
+    String encodedString = new String(responseData);
+    LOG.fine("tmaster returns physical plan in base64 str: " + encodedString);
+    // base64 string to proto bytes
+    byte[] decodedBytes = Base64.getDecoder().decode(encodedString);
+    // construct proto obj from bytes
+    PhysicalPlan physicalPlan = null;
+    try {
+      physicalPlan = PhysicalPlan.parseFrom(decodedBytes);
+    } catch (Exception e) {
+      throw new InvalidStateException(topologyName, "Failed to fetch the physical plan");
+    }
+    return physicalPlan;
   }
 
   @Override
   public synchronized PhysicalPlan get() {
-    physicalPlan = stateManagerAdaptor.getPhysicalPlan(topologyName);
-    if (physicalPlan == null) {
-      throw new InvalidStateException(topologyName, "Failed to fetch the physical plan");
+    TopologyMaster.TMasterLocation tMasterLocation
+        = stateManagerAdaptor.getTMasterLocation(topologyName);
+    String host = tMasterLocation.getHost();
+    int port = tMasterLocation.getControllerPort();
+
+    // construct metric cache stat url
+    String url = "http://" + host + ":" + port + "/get_current_physical_plan";
+    LOG.fine("tmaster physical plan query endpoint: " + url);
+
+    // http communication
+    HttpURLConnection con = NetworkUtils.getHttpConnection(url);
+    NetworkUtils.sendHttpGetRequest(con);
+    byte[] responseData = NetworkUtils.readHttpResponse(con);
+
+    physicalPlan = ParseResponseToPhysicalPlan(responseData);
+    return physicalPlan;
+  }
+
+  /**
+   * try best effort to return a latest physical plan
+   * 1. refresh physical plan
+   * 2. if refreshing fails, return the last physical plan
+   * @return physical plan
+   */
+  public PhysicalPlan getPhysicalPlan() {
+    try {
+      get();
+    } catch (InvalidStateException e) {
+      if (physicalPlan == null) {
+        throw e;
+      }
     }
     return physicalPlan;
   }
 
+  /**
+   * A utility method to extract bolt component names from the topology.
+   *
+   * @return list of all bolt names
+   */
+  public List<String> getBoltNames(PhysicalPlan pp) {
+    TopologyAPI.Topology localTopology = pp.getTopology();
+    ArrayList<String> boltNames = new ArrayList<>();
+    for (TopologyAPI.Bolt bolt : localTopology.getBoltsList()) {
+      boltNames.add(bolt.getComp().getName());
```

Comment:
getBoltsList() returns logical components or physical instances? I thought it is instances then we need to dedup here.